crates/biome_js_analyze/src/utils/restricted_glob.rs

```diff
@@ -98,12 +135,59 @@ impl schemars::JsonSchema for RestrictedGlob {
 /// Constructing candidates has a very small cost associated with it, so
 /// callers may find it beneficial to amortize that cost when matching a single
 /// path against multiple globs or sets of globs.
+#[derive(Debug, Clone)]
 pub struct CandidatePath<'a>(globset::Candidate<'a>);
 impl<'a> CandidatePath<'a> {
     /// Create a new candidate for matching from the given path.
     pub fn new(path: &'a impl AsRef<std::path::Path>) -> Self {
         Self(globset::Candidate::new(path))
     }
+
+    /// Tests whether the current path matches `glob`.
+    pub fn matches(&self, glob: &RestrictedGlob) -> bool {
+        glob.is_match_candidate(self)
+    }
+
+    /// Match against a list of globs where negated globs are handled as exceptions.
+    ///
+    /// Let's take an example:
+    ///
+    /// ```
+    /// use biome_js_analyze::utils::restricted_glob::{CandidatePath, RestrictedGlob};
+    ///
+    /// let globs: &[RestrictedGlob] = &[
+    ///     "*".parse().unwrap(),
+    ///     "!a*".parse().unwrap(),
+    ///     "a".parse().unwrap(),
+    /// ];
+    ///
+    /// assert!(CandidatePath::new(&"b").matches_with_exceptions(globs));
+    /// assert!(CandidatePath::new(&"a").matches_with_exceptions(globs));
+    ///
+    /// assert!(!CandidatePath::new(&"abc").matches_with_exceptions(globs));
+    /// ```
+    ///
+    /// - `b` matches `*` and is not excluded by the exception `!a*`.
+    ///   Thus, `b` matches the list of globs.
+    /// - `abc` matches the first glob `*`, however it is excluded by the exception `!a*`.
+    ///   Thus `abc` doesn't match the list of globs.
+    /// - `a` matches the first glob `*` and is excluded by the exception `!a*`.
+    ///   However, it is included again by the last glob `a`.
+    ///   Thus `a` matches the list of globs.
```

Comment:
Thank you for providing a very good explanation. 

Damn, these globs can create a lot confusion, based on this explanation. I don't know how to feel about them, because they have too much power. 

For example, `["a*", "!a", "a*", "!a*", "a"]`, which is an absurd edge case, means that the input glob `a` will match the last `a`. Does that mean that we need to check **all** globs? Or is that why we use `rev()`? 

Then I hope we actually explain that in our docs, and tell the users that the last glob is the one that wins.

Reply:
There are two ways of implementing it:

- From left to right (aka in-order)
  1. Looks for the first non-negated glob that matches
  2. From this first matching glob, looks for the first negated glob that doesn't match
  3. repeat
  4. If we end in step (1) then we have a match, otherwise we have no match
- From right to left (aka in reverse order)
  Looks for the first non-negated glob that matches or the first negated glob that matches
  If it is a non-negated glob, we have a match, otherwise we have no match

I choose the reverse order because it looks simpler to me and could avoid some glob matching.

You edge case that you present has the intended behavior. `.gitignore` has the same behavior.

Do you think I should add this edge case in the docs?

Reply:
Thank you for the explanation. 

> Do you think I should add this edge case in the docs?

Yes, but in the user docs when we teach them how to use the options :)